halo2_gadgets/src/poseidon/pow5.rs

```diff
@@ -94,20 +96,23 @@ impl<F: FieldExt, const WIDTH: usize, const RATE: usize> Pow5Chip<F, WIDTH, RATE
         meta.create_gate("full round", |meta| {
             let s_full = meta.query_selector(s_full);
 
-            (0..WIDTH)
-                .map(|next_idx| {
-                    let state_next = meta.query_advice(state[next_idx], Rotation::next());
-                    let expr = (0..WIDTH)
-                        .map(|idx| {
-                            let state_cur = meta.query_advice(state[idx], Rotation::cur());
-                            let rc_a = meta.query_fixed(rc_a[idx], Rotation::cur());
-                            pow_5(state_cur + rc_a) * m_reg[next_idx][idx]
-                        })
-                        .reduce(|acc, term| acc + term)
-                        .expect("WIDTH > 0");
-                    s_full.clone() * (expr - state_next)
-                })
-                .collect::<Vec<_>>()
+            Constraints::with_selector(
+                s_full,
+                (0..WIDTH)
+                    .map(|next_idx| {
+                        let state_next = meta.query_advice(state[next_idx], Rotation::next());
+                        let expr = (0..WIDTH)
+                            .map(|idx| {
+                                let state_cur = meta.query_advice(state[idx], Rotation::cur());
+                                let rc_a = meta.query_fixed(rc_a[idx], Rotation::cur());
+                                pow_5(state_cur + rc_a) * m_reg[next_idx][idx]
+                            })
+                            .reduce(|acc, term| acc + term)
+                            .expect("WIDTH > 0");
+                        expr - state_next
+                    })
+                    .collect::<Vec<_>>(),
```

Comment:
Collecting into a Vec shouldn't be necessary here, I think.
```suggestion
```

Reply:
It is necessary due to lifetimes. Believe me, I tried 🙂